File: test/pattern/test_template_method.cpp
#include <gtest/gtest.h>
#include "pattern/template_method.hpp"

using namespace pattern;

// ---------------------------------------------------------------------------
// CsvMiner
// ---------------------------------------------------------------------------
TEST(TemplateMethod, CsvMinerParsesRows)
{
    CsvMiner miner;
    miner.mine("a;b;c");
    EXPECT_EQ(3u, miner.parsedLines().size());
}

TEST(TemplateMethod, CsvMinerParsedLinesHavePrefix)
{
    CsvMiner miner;
    miner.mine("hello;world");
    EXPECT_EQ("csv:hello", miner.parsedLines()[0]);
    EXPECT_EQ("csv:world", miner.parsedLines()[1]);
}

TEST(TemplateMethod, CsvMinerReportContainsRowCount)
{
    CsvMiner miner;
    std::string report = miner.mine("a;b;c");
    EXPECT_NE(std::string::npos, report.find("3"));
}

TEST(TemplateMethod, CsvMinerReportContainsAnalysisNote)
{
    CsvMiner miner;
    std::string report = miner.mine("x;y");
    EXPECT_NE(std::string::npos, report.find("CSV analysis"));
}

TEST(TemplateMethod, CsvMinerSingleLine)
{
    CsvMiner miner;
    miner.mine("only");
    EXPECT_EQ(1u, miner.parsedLines().size());
    EXPECT_EQ("csv:only", miner.parsedLines()[0]);
}

// ---------------------------------------------------------------------------
// JsonMiner
// ---------------------------------------------------------------------------
TEST(TemplateMethod, JsonMinerParsesOneRecord)
{
    JsonMiner miner;
    miner.mine("{\"key\":\"value\"}");
    EXPECT_EQ(1u, miner.parsedLines().size());
}

TEST(TemplateMethod, JsonMinerParsedLineHasPrefix)
{
    JsonMiner miner;
    miner.mine("doc");
    EXPECT_EQ("json:doc", miner.parsedLines()[0]);
}

TEST(TemplateMethod, JsonMinerReportUsesDefaultAnalysis)
{
    JsonMiner miner;
    std::string report = miner.mine("{}");
    EXPECT_NE(std::string::npos, report.find("default analysis"));
}

// ---------------------------------------------------------------------------
// Template method is final — the skeleton is fixed
// ---------------------------------------------------------------------------
TEST(TemplateMethod, TemplateMethodAlwaysReturnsReport)
{
    CsvMiner  csv;
    JsonMiner json;
    EXPECT_FALSE(csv.mine("a").empty());
    EXPECT_FALSE(json.mine("b").empty());
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: scripts/test_pattern/test_template_method.py
"""Tests for the Template Method pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from template_method import CsvMiner, JsonMiner


class TestCsvMiner(unittest.TestCase):

    def test_parses_rows(self):
        miner = CsvMiner()
        miner.mine("a;b;c")
        self.assertEqual(3, len(miner.parsed_lines))

    def test_parsed_lines_have_prefix(self):
        miner = CsvMiner()
        miner.mine("hello;world")
        self.assertEqual("csv:hello", miner.parsed_lines[0])
        self.assertEqual("csv:world", miner.parsed_lines[1])

    def test_report_contains_row_count(self):
        miner  = CsvMiner()
        report = miner.mine("a;b;c")
        self.assertIn("3", report)

    def test_report_contains_analysis_note(self):
        miner  = CsvMiner()
        report = miner.mine("x;y")
        self.assertIn("CSV analysis", report)

    def test_single_line(self):
        miner = CsvMiner()
        miner.mine("only")
        self.assertEqual(1, len(miner.parsed_lines))
        self.assertEqual("csv:only", miner.parsed_lines[0])


class TestJsonMiner(unittest.TestCase):

    def test_parses_one_record(self):
        miner = JsonMiner()
        miner.mine('{"key":"value"}')
        self.assertEqual(1, len(miner.parsed_lines))

    def test_parsed_line_has_prefix(self):
        miner = JsonMiner()
        miner.mine("doc")
        self.assertEqual("json:doc", miner.parsed_lines[0])

    def test_report_uses_default_analysis(self):
        miner  = JsonMiner()
        report = miner.mine("{}")
        self.assertIn("default analysis", report)


class TestTemplateMethod(unittest.TestCase):

    def test_mine_always_returns_nonempty_report(self):
        self.assertTrue(CsvMiner().mine("a"))
        self.assertTrue(JsonMiner().mine("b"))


if __name__ == "__main__":
    unittest.main()
